### PR TITLE
Some modification and new files for QHA

### PR DESCRIPTION
**Pull request by samarerstm**
Some modification on qha.py
New file for  v-ZSISA-QHA approximation method ["Physical Review B 110 (1), 014103"]
New file for generating necessary deformation needed for ZSISA-QHA method

### CODE REVIEW

**Review by gpetretto — COMMENTED** [review]
Hi @samarerstm, sorry for the delay in reviewing this PR.
I went through the code and left comments mainly on the first two files. 

I did not go deep into the `QHA_App` because it is unclear to me how this should be integrated with the rest. Is the idea to keep it as an alternative and separate object with respect to the original `AbstractQHA`? Or do you plan to integrate it?
It seems that there is quite a bit of code duplication between `QHA_App` and `AbstractQHA`, so I was wondering if it wouldn't be possible to have different subclasses of the `AbstractQHA` to calculate different approximations (e.g. one subclass for E2Vib1, one for EinfVib1, and so on). More or less like now there are the `QHA`, `QHA3PF` and `QHA3P` subclasses to deal with other different approaches to the calculation. Or is this completely different?

Concerning the formatting of the code, I know that @gmatteo is not strict about PEP8 and similar coding conventions, but I think that it could be beneficial to pass the code through some tool like "black" to improve the formatting.

I should also add that I did not have the time to dig into the details of your paper, so I did check anything about the math or the formulas.

File: abipy/dfpt/deformation_utils.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
+# deformation_utils.py
+
+import numpy as np
+from pymatgen.core import Structure, Lattice, Element
+from abipy.core.symmetries import AbinitSpaceGroup
+from pymatgen.symmetry.analyzer import SpacegroupAnalyzer
+import re
+
+
+def generate_deformations_volumic(structure, eps_V=0.02, scales=[-1, 0, 1, 2, 3]):
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/deformation_utils.py]: Done

File: abipy/dfpt/deformation_utils.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    spgrp = AbinitSpaceGroup.from_structure(structure)
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/deformation_utils.py]: It is deleted

File: abipy/dfpt/deformation_utils.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        rprim2 = np.copy(rprim)
+        rprim2[:, :] = rprim[:, :] * (1.00 + eps_V * i)**(1/3.)
+        namei = int(round(1000 * (1.00 + eps_V * i)))
+        formatted_namei = f"{namei:04d}"
+
+        structure2 = structure.copy()
+        structure2.lattice = Lattice(rprim2)
+        structures_new[formatted_namei] = structure2
+        print(formatted_namei)
+        print(rprim2)
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/deformation_utils.py]: They are removed

File: abipy/dfpt/deformation_utils.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    structures_new = {}
+
+    for i in scales:
+        rprim2 = np.copy(rprim)
+        rprim2[:, :] = rprim[:, :] * (1.00 + eps_V * i)**(1/3.)
+        namei = int(round(1000 * (1.00 + eps_V * i)))
+        formatted_namei = f"{namei:04d}"
+
+        structure2 = structure.copy()
+        structure2.lattice = Lattice(rprim2)
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/deformation_utils.py:22]: This new command does not work.

File: abipy/dfpt/deformation_utils.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        structures_new[formatted_namei] = structure2
+        print(formatted_namei)
+        print(rprim2)
+
+    return structures_new
+
+def generate_deformations(structure , eps=0.005):
+    spgrp = AbinitSpaceGroup.from_structure(structure )
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/deformation_utils.py:32]: I attempt to use AbinitInput.abiget_spacegroup, but I have not succeed . could you please change it as your prefer ?

File: abipy/dfpt/qha.py [hunk 6]
```diff
@@ -173,38 +216,71 @@ def plot_energies(self, tstart=0, tstop=800, num=10, ax=None, **kwargs) -> Figur
...
 
         Returns: |Function1D|
         """
         f = self.fit_energies(tstart, tstop, num)
+        if (tref!= None):
+            f0 = self.fit_energies(tref, tref, 1)
 
         dt = f.temp[1] - f.temp[0]
-        alpha = (f.min_vol[2:] - f.min_vol[:-2]) / (2 * dt) / f.min_vol[1:-1]
+        if (self.eos_name == 'vinet') :
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/qha.py]: Now this method will be work for 
eos_list = [ 'taylor', 'murnaghan', 'birch', 'birch_murnaghan','pourier_tarantola', 'vinet', 'antonschmidt']
there will be 2 other eos "deltafactor" which is smooth itself and do not need this method and 
'numerical_eos' which does not work correctly for QHA method. most of my test with 'numerical_eos' failed and this eos can not interpolate the energies for my case.

File: abipy/dfpt/qha.py [hunk 7]
```diff
@@ -260,6 +336,177 @@ def plot_vol_vs_t(self, tstart=0, tstop=800, num=100, ax=None, **kwargs) -> Figu
...
+            num: int, optional Number of samples to generate. Default is 100.
+            ax: |matplotlib-Axes| or None if a new figure should be created.
+
+        Returns: |matplotlib-Figure|
+        """
+        f =  self.fit_energies(tstart, tstop, num)
+        param = np.zeros((num,4))
+        param=np.polyfit(self.volumes, self.angles_alpha, 3)
+        pa = np.poly1d(param)
+        gamma=pa(f.min_vol)
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/qha.py]: Done

File: abipy/dfpt/qha.py [hunk 8]
```diff
@@ -260,6 +336,177 @@ def plot_vol_vs_t(self, tstart=0, tstop=800, num=100, ax=None, **kwargs) -> Figu
...
+        param=np.polyfit(self.volumes, self.angles_beta , 3)
+        pb = np.poly1d(param)
+        beta=pb(f.min_vol)
+        param=np.polyfit(self.volumes, self.angles_gama , 3)
+        pc = np.poly1d(param)
+        alpha=pc(f.min_vol)
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/qha.py]: Done

File: abipy/dfpt/qha.py [hunk 9]
```diff
@@ -260,6 +336,177 @@ def plot_vol_vs_t(self, tstart=0, tstop=800, num=100, ax=None, **kwargs) -> Figu
...
+
+        aa_qha,bb_qha,cc_qha = self.get_abc(tstart, tstop, num)
+        if (tref!=None):
+            aa_tref,bb_tref,cc_tref = self.get_abc(tref, tref, 1)
+
+        alpha_qha_a = np.zeros( num-2)
+        alpha_qha_b = np.zeros( num-2)
+        alpha_qha_c = np.zeros( num-2)
+        if (tref==None):
+            alpha_qha_a = (aa_qha[2:] - aa_qha[:-2]) / (2 * dt) / aa_qha[1:-1]
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/qha.py]: In new version plot_thermal_expansion_coeff_abc in inconsistent with plot_thermal_expansion_coeff

File: abipy/dfpt/deformation_utils.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    structures_new = {}
+
+    if 1 <= spgrp_number <= 2:
+        disp=[[1,1,1,1,1,1],  [0,1,1,1,1,1],  [2,1,1,1,1,1],  [1,0,1,1,1,1],  [1,2,1,1,1,1],  [1,1,0,1,1,1],  
+              [1,1,2,1,1,1],  [1,1,1,0,1,1],  [1,1,1,2,1,1],  [1,1,1,1,0,1],  [1,1,1,1,2,1],  [1,1,1,1,1,0],  
+              [1,1,1,1,1,2],  [0,0,1,1,1,1],  [1,0,0,1,1,1],  [1,1,0,0,1,1],  [1,1,1,0,0,1],  [1,1,1,1,0,0],  
+              [0,1,0,1,1,1],  [0,1,1,0,1,1],  [0,1,1,1,0,1],  [0,1,1,1,1,0],  [1,0,1,0,1,1],  [1,0,1,1,0,1],  
+              [1,0,1,1,1,0],  [1,1,0,1,0,1],  [1,1,0,1,1,0],  [1,1,1,0,1,0] , [0 ,0,0,0,0,0]]  
+        #if abs(rprim[1, 0]) > 1e-9 or abs(rprim[2, 0]) > 1e-9 or abs(rprim[2, 1]) > 1e-9:
+        print("Warning: The lattice is oriented such that xz =xy =yz =0 .")
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/deformation_utils.py]: it is corrected

File: abipy/dfpt/deformation_utils.py [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+
+            namei = int(round(1000 * (1.00 + eps * i)))
+            namej = int(round(1000 * (1.00 + eps * j)))
+            namek = int(round(1000 * (1.00 + eps * k)))
+            namel = int(round(1000 * (1.00 + eps * l)))
+            namem = int(round(1000 * (1.00 + eps * m)))
+            namen = int(round(1000 * (1.00 + eps * n)))
+            formatted_namei = f"{namei:04d}_{namej:04d}_{namek:04d}_{namel:04d}_{namem:04d}_{namen:04d}"
+
+            structure2=structure
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/deformation_utils.py]: corrected

File: abipy/dfpt/qha.py [hunk 12]
```diff
@@ -94,6 +102,40 @@ def fit_energies(self, tstart=0, tstop=800, num=100):
...
+        fits = [self.eos.fit(self.volumes, e) for e in tot_en.T]
+
+        # list of minimum volumes and energies, one for each temperature
+        v0 = np.array([fit.v0 for fit in fits])
+        e0 = np.array([fit.e0 for fit in fits])
+        b0 = np.array([fit.b0 for fit in fits])
+        b1 = np.array([fit.b1 for fit in fits])
+
+        #print (b0,b1,e0,v0)
+        v = np.array([fit.v0 for fit in fits])
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/qha.py]: Thanks for this point. In new version it is corrected

File: abipy/dfpt/qha.py [hunk 13]
```diff
@@ -94,6 +102,40 @@ def fit_energies(self, tstart=0, tstop=800, num=100):
...
+
+        Returns:
+            F2D: Second derivative of free energy with respect to volume, based on the Vinet equation. 
+        """
+        tmesh = np.linspace(tstart, tstop, num)
+
+        # array with phonon energies and shape (n_vol, n_temp)
+
+        ph_energies = self.get_vib_free_energies(tstart, tstop, num)
+        tot_en = self.energies[np.newaxis, :].T + ph_energies
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/qha.py]: it is corrected.
Actually I do not find how to   adjust this pressure  . it is not an input of any functions in QHA class or not read from any input.

File: abipy/dfpt/qha.py [hunk 14]
```diff
@@ -94,6 +102,40 @@ def fit_energies(self, tstart=0, tstop=800, num=100):
...
+        """
+        Performs a fit of the energies as a function of the volume at different temperatures.
+
+        Args:
+            tstart: The starting value (in Kelvin) of the temperature mesh.
+            tstop: The end value (in Kelvin) of the mesh.
+            num: int, optional Number of samples to generate. Default is 100.
+
+        Returns:
+            F2D: Second derivative of free energy with respect to volume, based on the Vinet equation.
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/qha.py]: Now it is compatible with eos_list methods

File: abipy/dfpt/qha.py [hunk 15]
```diff
@@ -94,6 +102,40 @@ def fit_energies(self, tstart=0, tstop=800, num=100):
         min_energies = np.array([fit.e0 for fit in fits])
 
         return dict2namedtuple(tot_en=tot_en, fits=fits, min_en=min_energies, min_vol=min_volumes, temp=tmesh)
+    def fit_free_energies(self, tstart, tstop, num):
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/qha.py]: I merged this function with fit_free_energies

File: abipy/dfpt/qha.py [hunk 16]
```diff
@@ -173,38 +216,71 @@ def plot_energies(self, tstart=0, tstop=800, num=10, ax=None, **kwargs) -> Figur
...
             tstop: The end value (in Kelvin) of the mesh.
+            tref: The reference temperature (in Kelvin) used to compute the thermal expansion coefficient 1/V(tref) * dV(T)/dT.
+                  (If tref is not available, it uses 1/V(T) * dV(T)/dT instead.)
+
             num: int, optional Number of samples to generate. Default is 100.
 
         Returns: |Function1D|
         """
         f = self.fit_energies(tstart, tstop, num)
+        if (tref!= None):
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/qha.py]: Done

File: abipy/dfpt/qha_aproximation.py [hunk 17]
```diff
@@ -0,0 +1,1971 @@
...
+from pymatgen.analysis.eos import EOS
+from abipy.core.func1d import Function1D
+from abipy.tools.plotting import add_fig_kwargs, get_ax_fig_plt, get_axarray_fig_plt
+from abipy.electrons.gsr import GsrFile
+from abipy.dfpt.ddb import DdbFile
+from abipy.dfpt.phonons import PhononBandsPlotter, PhononDos, PhdosFile
+from abipy.dfpt.gruneisen import GrunsNcFile
+
+
+class QHA_App(metaclass=abc.ABCMeta):
```
[the remark this answers is not in the document]
Comment (samarerstm) [reply abipy/dfpt/qha_aproximation.py:20]: I need to discuss this file with Matteo. He asked me to submit a merge request so he can review it.

**Pull request merged**